lib/matplotlib/axes/_axes.py

```diff
@@ -4742,64 +4742,86 @@ def fill(self, *args, **kwargs):
+            ``x[where]``.  More precisely, fill between ``x[i]`` and ``x[i+1]``
+            if ``where[i] and where[i+1]``.  Note that this definition implies
+            that an isolated *True* value between two *False* values in
+            *where* will not result in filling.  Both sides of the *True*
+            position remain unfilled due to the adjacent *False* values.
 
         interpolate : bool, optional
-            If `True`, interpolate between the two lines to find the
-            precise point of intersection.  Otherwise, the start and
-            end points of the filled region will only occur on explicit
-            values in the *x* array.
+            This option is only relvant if *where* is used and the two curves
+            are crossing each other.
 
-        step : {'pre', 'post', 'mid'}, optional
-            If not None, fill with step logic.
+            Semantically, *where* is often used for *y1* > *y2* or similar.
+            By default, the nodes of the polygon defining the filled region
+            will only be placed at the positions in the *x* array.  Such a
+            polygon cannot describe the above semantics close to the
+            intersection.  The x-sections containing the intersecion are
+            simply clipped.
 
-        Returns
-        -------
-        `PolyCollection`
-            Plotted polygon collection
+            Setting *interpolate* to *True* will calculate the actual
+            interscection point and extend the filled region up to this point.
 
-        Notes
-        -----
+        step : {'pre', 'post', 'mid'}, optional
+            Define *step* if the filling should be a step function,
+            i.e. constant in between *x*. The value determines where the
+            step will occur:
 
-        Additional Keyword args passed on to the
-        :class:`~matplotlib.collections.PolyCollection`.
+            - 'pre': The y value is continued constantly to the left from
+              every *x* position.
+            - 'post': The y value is continued constantly to the right from
+              every *x* position.
+            - 'mid': Steps occur in the middle between the *x* positions.
 
-        kwargs control the :class:`~matplotlib.patches.Polygon` properties:
+        Other Parameters
+        ----------------
+        **kwargs
+            All other keyword arguments are passed on to `~.PolyCollection`.
+            They control the `~.Polygon` properties:
 
-        %(PolyCollection)s
+            %(PolyCollection)s
+
+        Returns
+        -------
+        `~.PolyCollection`
+            A `~.PolyCollection` containing the plotted polygons.
 
         See Also
         --------
+        fill_betweenx : Fill between two sets of x-values.
 
-            :meth:`fill_betweenx`
-                for filling between two sets of x-values
+        Notes
+        -----
+        .. [notes section required to get data note injection right]
```

Comment:
This seems like a bug in the data decorator...

Reply:
I've summarized the problem and possible solutions (of which the above is one) in #10189.